netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -150,18 +153,27 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
    */
   @Override
   public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-    if (msg instanceof CreateStreamCommand) {
-      createStream((CreateStreamCommand) msg, promise);
-    } else if (msg instanceof SendGrpcFrameCommand) {
-      sendGrpcFrame(ctx, (SendGrpcFrameCommand) msg, promise);
-    } else if (msg instanceof CancelStreamCommand) {
-      cancelStream(ctx, (CancelStreamCommand) msg, promise);
-    } else if (msg instanceof RequestMessagesCommand) {
-      ((RequestMessagesCommand) msg).requestMessages();
-    } else if (msg instanceof SendPingCommand) {
-      sendPingFrame(ctx, (SendPingCommand) msg, promise);
-    } else {
-      throw new AssertionError("Write called for unexpected type: " + msg.getClass().getName());
+    try {
+      if (msg instanceof CreateStreamCommand) {
+        createStream((CreateStreamCommand) msg, promise);
+      } else if (msg instanceof SendGrpcFrameCommand) {
+        sendGrpcFrame(ctx, (SendGrpcFrameCommand) msg, promise);
+      } else if (msg instanceof CancelStreamCommand) {
+        cancelStream(ctx, (CancelStreamCommand) msg, promise);
+      } else if (msg instanceof RequestMessagesCommand) {
+        ((RequestMessagesCommand) msg).requestMessages();
+      } else if (msg instanceof SendPingCommand) {
+        sendPingFrame(ctx, (SendPingCommand) msg, promise);
+      } else {
+        throw new AssertionError("Write called for unexpected type: " + msg.getClass().getName());
+      }
+    } catch (StatusException e) {
+      promise.setFailure(e);
+      try {
+        super.close(ctx, ctx.newPromise());
+      } catch (Exception ce) {
```

Comment:
Writes in Netty do not typically throw. Instead they fail the promise with the exception.  We should be following that pattern.

I'll revert the code to the previous version.